homeassistant/components/sensor/mqtt.py

```diff
@@ -104,6 +111,20 @@ def message_received(topic, payload, qos):
                 self._expiration_trigger = async_track_point_in_utc_time(
                     self.hass, self.value_is_expired, expiration_at)
 
+            if len(self._json_attributes) > 0:
+                self._attributes = set({})
```

Comment:
I think there was a misunderstanding: `self._json_attributes` (not `self._attributes`) should be made a set, so that the computation of `self._json_attributes & json_dict.keys()` is efficient (intersection of two ordered iterables should be linear…)

Making `self._attributes` a set instead a dict here will provoke errors when `self._json_attributes & json_dict.keys()` is empty, because then `device_state_attributes()` returns a set instead of a dict. Please add a test that tests this case, i.e., when the set of selected attributes and the set of keys in the JSON payload are disjunct.

Reply:
Sorry, I am no python Pro. I have had another go and trying to make self._json_attributes a set.

I have also added another couple of tests based on REST json attribute commit